url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/model/Model.java

```diff
@@ -0,0 +1,334 @@
+        boolean[][][] shapeMatrix = shape.getMatrix();
+        boolean band = false;
+        for (int k = 0; k < shapeMatrix[0][0].length; ++k) {
+            int minY = 0;
+            for (int i = 0; i < shapeMatrix.length; ++i) {
+                band = false;
+                for (int j = 0; j < shapeMatrix[i].length; ++j) {
+                    if (shapeMatrix[i][j][k]) {
+                        band = true;
+                        minY = Math.min(minY, j);
+                    }
+                }
+                if (band
+                        && matrix[i + shape.getDesfX()][minY + shape.getDesfY()
+                        - 1][k + shape.getLevel()]) {
+                    return;
+                }
+            }
+        }
+        shape.decrementDesfY();
+    }
+
+    private boolean rotate(Shape shape, int rotation) {
+        shape.rotate(rotation);
+        int x, y, z;
+        boolean[][][] shapeMatrix = shape.getMatrix();
+
+        for (int i = 0; i < shapeMatrix.length; ++i) {
+            for (int j = 0; j < shapeMatrix[i].length; ++j) {
+                for (int k = 0; k < shapeMatrix[i][j].length; ++k) {
+                    x = i + shape.getDesfX();
+                    y = j + shape.getDesfY();
+                    z = k + shape.getLevel();
+                    if (x < matrix.length && y < matrix[0].length && z < matrix[0][0].length) {
+                        if (matrix[x][y][z] && shapeMatrix[i][j][k]) {
+                            for (int p = 0; p < 3; ++p) {
+                                shape.rotate(rotation);
+                            }
+                            return false;
+                        }
+                    }
+
+                }
+            }
+        }
+
+        return true;
+    }
+
+    public boolean rotateYes(Shape shape) {
+        return rotate(shape, YES);
+    }
+
+    private boolean rotateReverse(Shape shape, int rotation) {
+        if (rotate(shape, rotation)) {
+            if (rotate(shape, rotation)) {
+                if (rotate(shape, rotation)) {
+                    return true;
+                } else {
+                    shape.rotate(rotation);
+                    shape.rotate(rotation);
+                }
+            } else {
+                shape.rotate(rotation);
+                shape.rotate(rotation);
+                shape.rotate(rotation);
+            }
+        }
+        return false;
+    }
+
+    public boolean rotateYesReverse(Shape shape) {
+        return rotateReverse(shape, YES);
+    }
+
+    public boolean rotateNo(Shape shape) {
+        return rotate(shape, NO);
+    }
+
+    public boolean rotateNoReverse(Shape shape) {
+        return rotateReverse(shape, NO);
+    }
+
+    public boolean rotateMaybe(Shape shape) {
+        return rotate(shape, MAYBE);
+    }
+
+    public boolean rotateMaybeReverse(Shape shape) {
+        return rotateReverse(shape, MAYBE);
+    }
+
+    public boolean incrementLevel(Shape shape) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 253-265 into a separate method.